File: process.go
package clauderelay

import (
	"bytes"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

type ConversationState struct {
	History   []string
	Timestamp time.Time
}


type ClaudeProcess struct {
	cmd              *exec.Cmd
	pty              *os.File
	tempDir          string
	conversationHistory []string
	conversationStates  []ConversationState
	lastUndoneHistory   []string  // Store the last undone conversation for restore
	setup            *ClaudeSetup
}

func NewClaudeProcess(setup *ClaudeSetup) (*ClaudeProcess, error) {
	// First ensure we have a config file to skip welcome
	configDir := filepath.Join(setup.GetClaudeHome(), ".config", "claude")
	os.MkdirAll(configDir, 0755)
	
	configFile := filepath.Join(configDir, "config.json")
	if _, err := os.Stat(configFile); os.IsNotExist(err) {
		// Create config with theme to skip welcome
		config := `{"theme":"dark","outputStyle":"default"}`
		os.WriteFile(configFile, []byte(config), 0644)
	}

	tempDir, err := ioutil.TempDir("", "claude-relay-*")
	if err != nil {
		return nil, fmt.Errorf("failed to create temp directory: %w", err)
	}

	// We don't actually need to start a persistent process
	// since we use --print mode for each message
	// Just return the structure
	return &ClaudeProcess{
		cmd:              nil,
		pty:              nil,
		tempDir:          tempDir,
		conversationHistory: []string{},
		conversationStates:  []ConversationState{},
		lastUndoneHistory:   nil,
		setup:            setup,
	}, nil
}

func (p *ClaudeProcess) Kill() error {
	if p.cmd != nil && p.cmd.Process != nil {
		log.Printf("Killing process PID: %d", p.cmd.Process.Pid)
		if err := p.cmd.Process.Kill(); err != nil {
			return fmt.Errorf("failed to kill process: %w", err)
		}
		p.cmd.Wait()
	}
	return nil
}

func (p *ClaudeProcess) Cleanup() error {
	if p.pty != nil {
		p.pty.Close()
	}

	if p.tempDir != "" {
		if err := os.RemoveAll(p.tempDir); err != nil {
			return fmt.Errorf("failed to remove temp directory: %w", err)
		}
	}

	return nil
}

func (p *ClaudeProcess) IsRunning() bool {
	if p.cmd == nil || p.cmd.Process == nil {
		return false
	}
	
	process, err := os.FindProcess(p.cmd.Process.Pid)
	if err != nil {
		return false
	}
	
	err = process.Signal(os.Signal(nil))
	return err == nil
}

func (p *ClaudeProcess) GetWorkingDirectory() string {
	return p.tempDir
}

func (p *ClaudeProcess) SaveFile(filename string, content []byte) error {
	if p.tempDir == "" {
		return fmt.Errorf("no temporary directory available")
	}
	
	filePath := filepath.Join(p.tempDir, filename)
	return ioutil.WriteFile(filePath, content, 0644)
}

func (p *ClaudeProcess) ReadFile(filename string) ([]byte, error) {
	if p.tempDir == "" {
		return nil, fmt.Errorf("no temporary directory available")
	}
	
	filePath := filepath.Join(p.tempDir, filename)
	return ioutil.ReadFile(filePath)
}

func (p *ClaudeProcess) SendMessage(message string) (string, error) {
	// Add user message to history
	p.conversationHistory = append(p.conversationHistory, "User: "+message)
	
	// Build context from conversation history
	var contextBuilder strings.Builder
	if len(p.conversationHistory) > 1 {
		contextBuilder.WriteString("Previous conversation:\n")
		for _, msg := range p.conversationHistory[:len(p.conversationHistory)-1] {
			contextBuilder.WriteString(msg + "\n")
		}
		contextBuilder.WriteString("\nLatest message: " + message)
	} else {
		contextBuilder.WriteString(message)
	}
	
	fullPrompt := contextBuilder.String()
	
	// Use claude --print mode for this single request WITHOUT PTY
	// PTY seems to interfere with --print mode's stdin handling
	cmd := exec.Command(p.setup.GetClaudePath(), "--print", "--dangerously-skip-permissions")
	// Run from the base directory for workspace isolation
	cmd.Dir = p.setup.GetBaseDir()
	// Use system environment to access authentication, but add Claude-specific env vars
	cmd.Env = os.Environ()
	cmd.Env = append(cmd.Env, 
		fmt.Sprintf("BUN_INSTALL=%s", filepath.Join(p.setup.GetBaseDir(), ".bun")),
		fmt.Sprintf("PATH=%s:%s", filepath.Join(p.setup.GetBaseDir(), ".bun", "bin"), os.Getenv("PATH")),
	)
	// Add additional environment for relay
	cmd.Env = append(cmd.Env,
		"CLAUDE_RELAY=true",
		"TERM=dumb",
		"NO_COLOR=1",
	)
	
	// Use regular pipes for --print mode
	cmd.Stdin = strings.NewReader(fullPrompt)
	
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr
	
	
	// Run the command
	err := cmd.Run()
	if err != nil {
		stderrStr := stderr.String()
		
		if p.setup.IsAuthenticationNeeded(stderrStr) {
			return "", fmt.Errorf("authentication required: please restart the server to login")
		}
		return "", fmt.Errorf("claude command failed: %w, stderr: %s", err, stderrStr)
	}
	
	responseStr := stdout.String()
	
	// Add Claude's response to history
	p.conversationHistory = append(p.conversationHistory, "Claude: "+responseStr)
	
	// Keep history manageable (last 10 exchanges)
	if len(p.conversationHistory) > 20 {
		p.conversationHistory = p.conversationHistory[2:]
	}
	
	return responseStr, nil
}

func (p *ClaudeProcess) SendMessageWithProgress(message string, progressChan chan<- []byte, doneChan <-chan bool) (string, error) {
	// Save current state before processing (for undo functionality)
	p.SaveState()
	
	// Add user message to history
	p.conversationHistory = append(p.conversationHistory, "User: "+message)
	
	// Build context from conversation history
	var contextBuilder strings.Builder
	if len(p.conversationHistory) > 1 {
		contextBuilder.WriteString("Previous conversation:\n")
		for _, msg := range p.conversationHistory[:len(p.conversationHistory)-1] {
			contextBuilder.WriteString(msg + "\n")
		}
		contextBuilder.WriteString("\nLatest message: " + message)
	} else {
		contextBuilder.WriteString(message)
	}
	
	fullPrompt := contextBuilder.String()
	
	// Use claude --print mode for this single request
	cmd := exec.Command(p.setup.GetClaudePath(), "--print", "--dangerously-skip-permissions")
	// Run from the base directory for workspace isolation
	cmd.Dir = p.setup.GetBaseDir()
	// Use system environment to access authentication, but add Claude-specific env vars
	cmd.Env = os.Environ()
	cmd.Env = append(cmd.Env, 
		fmt.Sprintf("BUN_INSTALL=%s", filepath.Join(p.setup.GetBaseDir(), ".bun")),
		fmt.Sprintf("PATH=%s:%s", filepath.Join(p.setup.GetBaseDir(), ".bun", "bin"), os.Getenv("PATH")),
	)
	// Add additional environment for relay
	cmd.Env = append(cmd.Env,
		"CLAUDE_RELAY=true",
		"TERM=xterm",
		"NO_COLOR=1",
	)
	
	cmd.Stdin = strings.NewReader(fullPrompt)
	
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr
	
	// Start progress indicator in background
	progressTicker := make(chan struct{})
	go func() {
		messages := []string{
			"💭 Processing your request...",
			"🔍 Analyzing context...",
			"📖 Gathering information...",
			"🧠 Formulating response...",
		}
		messageIndex := 0
		
		ticker := time.NewTimer(2 * time.Second) // First update after 2 seconds
		defer ticker.Stop()
		
		for {
			select {
			case <-ticker.C:
				if messageIndex < len(messages) {
					select {
					case progressChan <- []byte(messages[messageIndex]):
					case <-doneChan:
						return
					}
					messageIndex++
					if messageIndex < len(messages) {
						ticker.Reset(3 * time.Second) // Subsequent updates every 3 seconds
					}
				}
			case <-progressTicker:
				return
			case <-doneChan:
				return
			}
		}
	}()
	
	// Run the command
	err := cmd.Run()
	close(progressTicker) // Stop progress updates
	
	if err != nil {
		// Check if authentication is needed
		stderrStr := stderr.String()
		
		if p.setup.IsAuthenticationNeeded(stderrStr) {
			return "", fmt.Errorf("authentication required: please restart the server to login")
		}
		return "", fmt.Errorf("claude command failed: %w, stderr: %s", err, stderrStr)
	}
	
	finalResponse := stdout.String()
	
	// Add Claude's response to history
	p.conversationHistory = append(p.conversationHistory, "Claude: "+finalResponse)
	
	// Keep history manageable (last 10 exchanges)
	if len(p.conversationHistory) > 20 {
		p.conversationHistory = p.conversationHistory[2:]
	}
	
	return finalResponse, nil
}

func (p *ClaudeProcess) SaveState() {
	// Make a copy of current conversation history
	historyCopy := make([]string, len(p.conversationHistory))
	copy(historyCopy, p.conversationHistory)
	
	state := ConversationState{
		History:   historyCopy,
		Timestamp: time.Now(),
	}
	
	p.conversationStates = append(p.conversationStates, state)
	
	// Keep only last 10 states to manage memory
	if len(p.conversationStates) > 10 {
		p.conversationStates = p.conversationStates[1:]
	}
}

func (p *ClaudeProcess) UndoLastExchange() error {
	if len(p.conversationStates) == 0 {
		return fmt.Errorf("no conversation states to undo")
	}
	
	// Get the last saved state
	lastState := p.conversationStates[len(p.conversationStates)-1]
	
	// Restore conversation history to that state
	p.conversationHistory = make([]string, len(lastState.History))
	copy(p.conversationHistory, lastState.History)
	
	// Remove the used state
	p.conversationStates = p.conversationStates[:len(p.conversationStates)-1]
	
	return nil
}

func (p *ClaudeProcess) UndoToIndex(messageIndex int) error {
	// Calculate which conversation history index this corresponds to
	// Each exchange has 2 entries (User: and Claude:)
	historyIndex := messageIndex * 2
	
	if historyIndex < 0 || historyIndex > len(p.conversationHistory) {
		return fmt.Errorf("invalid undo index: %d", messageIndex)
	}
	
	// Save the conversation that will be undone (for restore)
	if historyIndex < len(p.conversationHistory) {
		p.lastUndoneHistory = make([]string, len(p.conversationHistory))
		copy(p.lastUndoneHistory, p.conversationHistory)
	}
	
	// Truncate conversation history to this point
	p.conversationHistory = p.conversationHistory[:historyIndex]
	
	// Also truncate conversation states if needed
	// Find the appropriate state to restore
	for i := len(p.conversationStates) - 1; i >= 0; i-- {
		if len(p.conversationStates[i].History) <= historyIndex {
			p.conversationStates = p.conversationStates[:i+1]
			break
		}
	}
	
	return nil
}

func (p *ClaudeProcess) CanUndo() bool {
	return len(p.conversationStates) > 0
}

func (p *ClaudeProcess) GetLastExchange() (string, string, error) {
	if len(p.conversationHistory) < 2 {
		return "", "", fmt.Errorf("no complete exchange to return")
	}
	
	// Get last user message and Claude response
	userMsg := p.conversationHistory[len(p.conversationHistory)-2]
	claudeMsg := p.conversationHistory[len(p.conversationHistory)-1]
	
	// Strip prefixes
	userMsg = strings.TrimPrefix(userMsg, "User: ")
	claudeMsg = strings.TrimPrefix(claudeMsg, "Claude: ")
	
	return userMsg, claudeMsg, nil
}

func (p *ClaudeProcess) CanRestore() bool {
	return p.lastUndoneHistory != nil && len(p.lastUndoneHistory) > len(p.conversationHistory)
}

func (p *ClaudeProcess) RestoreLastUndo() ([]string, error) {
	if !p.CanRestore() {
		return nil, fmt.Errorf("nothing to restore")
	}
	
	// Get the messages that will be restored (for client display)
	restoredMessages := p.lastUndoneHistory[len(p.conversationHistory):]
	
	// Restore the conversation history
	p.conversationHistory = make([]string, len(p.lastUndoneHistory))
	copy(p.conversationHistory, p.lastUndoneHistory)
	
	// Clear the undo buffer since we've restored it
	p.lastUndoneHistory = nil
	
	// Rebuild conversation states
	p.SaveState()
	
	return restoredMessages, nil
}

func (p *ClaudeProcess) GetRestoredMessagesForClient() []map[string]string {
	if !p.CanRestore() {
		return nil
	}
	
	// Get messages that would be restored
	restoredPart := p.lastUndoneHistory[len(p.conversationHistory):]
	
	var messages []map[string]string
	for i := 0; i < len(restoredPart); i += 2 {
		if i+1 < len(restoredPart) {
			userMsg := strings.TrimPrefix(restoredPart[i], "User: ")
			claudeMsg := strings.TrimPrefix(restoredPart[i+1], "Claude: ")
			messages = append(messages, map[string]string{
				"user": userMsg,
				"claude": claudeMsg,
			})
		}
	}
	
	return messages
}

File: src/process.rs
use crate::error::{ClaudeRelayError, Result};
use crate::setup::ClaudeSetup;
use chrono::{DateTime, Utc};
use std::fs;
use std::io::Write;
use std::path::Path;
use std::process::{Command, Stdio};
use std::sync::Arc;
use tempfile::TempDir;

#[derive(Clone, Debug)]
pub struct ConversationState {
    pub history: Vec<String>,
    pub timestamp: DateTime<Utc>,
}

pub struct ClaudeProcess {
    temp_dir: TempDir,
    conversation_history: Vec<String>,
    conversation_states: Vec<ConversationState>,
    last_undone_history: Option<Vec<String>>,
    setup: Arc<ClaudeSetup>,
}

impl ClaudeProcess {
    pub fn new(setup: Arc<ClaudeSetup>) -> Result<Self> {
        // Ensure config file exists to skip welcome
        let config_dir = setup.get_claude_home().join(".config").join("claude");
        fs::create_dir_all(&config_dir)?;
        
        let config_file = config_dir.join("config.json");
        if !config_file.exists() {
            let config = r#"{"theme":"dark","outputStyle":"default"}"#;
            fs::write(&config_file, config)?;
        }

        let temp_dir = TempDir::new()
            .map_err(|e| ClaudeRelayError::Process(format!("Failed to create temp directory: {}", e)))?;

        Ok(ClaudeProcess {
            temp_dir,
            conversation_history: Vec::new(),
            conversation_states: Vec::new(),
            last_undone_history: None,
            setup,
        })
    }

    pub fn get_working_directory(&self) -> &Path {
        self.temp_dir.path()
    }

    pub fn save_file(&self, filename: &str, content: &[u8]) -> Result<()> {
        let file_path = self.temp_dir.path().join(filename);
        fs::write(&file_path, content)?;
        Ok(())
    }

    pub fn read_file(&self, filename: &str) -> Result<Vec<u8>> {
        let file_path = self.temp_dir.path().join(filename);
        let content = fs::read(&file_path)?;
        Ok(content)
    }

    pub fn send_message(&mut self, message: &str) -> Result<String> {
        // Add user message to history
        self.conversation_history.push(format!("User: {}", message));
        
        // Build context from conversation history
        let full_prompt = if self.conversation_history.len() > 1 {
            let mut context = String::from("Previous conversation:\n");
            for msg in &self.conversation_history[..self.conversation_history.len() - 1] {
                context.push_str(msg);
                context.push('\n');
            }
            context.push_str("\nLatest message: ");
            context.push_str(message);
            context
        } else {
            message.to_string()
        };
        
        // Use claude --print mode for this single request
        let mut cmd = Command::new(self.setup.get_claude_path());
        cmd.args(&["--print", "--dangerously-skip-permissions"])
            .current_dir(self.setup.get_base_dir())
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped());
        
        // Set environment
        for (key, value) in self.setup.get_claude_env() {
            cmd.env(key, value);
        }
        
        // Add additional environment for relay
        cmd.env("CLAUDE_RELAY", "true")
            .env("TERM", "dumb")
            .env("NO_COLOR", "1");
        
        let mut child = cmd.spawn()
            .map_err(|e| ClaudeRelayError::Process(format!("Failed to spawn Claude: {}", e)))?;
        
        // Write prompt to stdin
        if let Some(mut stdin) = child.stdin.take() {
            stdin.write_all(full_prompt.as_bytes())
                .map_err(|e| ClaudeRelayError::Process(format!("Failed to write to stdin: {}", e)))?;
        }
        
        let output = child.wait_with_output()
            .map_err(|e| ClaudeRelayError::Process(format!("Claude command failed: {}", e)))?;
        
        if !output.status.success() {
            let stderr = String::from_utf8_lossy(&output.stderr);
            
            if self.setup.is_authentication_needed(&stderr) {
                return Err(ClaudeRelayError::Authentication(
                    "Authentication required: please restart the server to login".into()
                ));
            }
            return Err(ClaudeRelayError::Process(
                format!("Claude command failed: {}", stderr)
            ));
        }
        
        let response = String::from_utf8_lossy(&output.stdout).to_string();
        
        // Add Claude's response to history
        self.conversation_history.push(format!("Claude: {}", response));
        
        // Keep history manageable (last 10 exchanges)
        if self.conversation_history.len() > 20 {
            self.conversation_history.drain(0..2);
        }
        
        Ok(response)
    }

    pub async fn send_message_with_progress<F>(
        &mut self,
        message: &str,
        _progress_callback: F,
    ) -> Result<String> 
    where
        F: FnMut(&str),
    {
        // Save current state before processing (for undo functionality)
        self.save_state();
        
        // Send progress updates
        let messages = [
            "💭 Processing your request...",
            "🔍 Analyzing context...",
            "📖 Gathering information...",
            "🧠 Formulating response...",
        ];
        
        // Start a task to send progress updates
        let (tx, mut rx) = tokio::sync::mpsc::channel(1);
        let progress_task = tokio::spawn(async move {
            let mut index = 0;
            let mut interval = tokio::time::interval(tokio::time::Duration::from_secs(2));
            
            loop {
                tokio::select! {
                    _ = interval.tick() => {
                        if index < messages.len() {
                            if tx.send(messages[index]).await.is_err() {
                                break;
                            }
                            index += 1;
                        }
                    }
                    _ = rx.recv() => {
                        break;
                    }
                }
            }
        });
        
        // Send the actual message
        let result = self.send_message(message);
        
        // Stop progress updates
        drop(progress_task);
        
        result
    }

    pub fn save_state(&mut self) {
        let state = ConversationState {
            history: self.conversation_history.clone(),
            timestamp: Utc::now(),
        };
        
        self.conversation_states.push(state);
        
        // Keep only last 10 states to manage memory
        if self.conversation_states.len() > 10 {
            self.conversation_states.remove(0);
        }
    }

    pub fn undo_last_exchange(&mut self) -> Result<()> {
        if self.conversation_states.is_empty() {
            return Err(ClaudeRelayError::Process("No conversation states to undo".into()));
        }
        
        // Get the last saved state
        let last_state = self.conversation_states.pop().unwrap();
        
        // Restore conversation history to that state
        self.conversation_history = last_state.history;
        
        Ok(())
    }

    pub fn undo_to_index(&mut self, message_index: usize) -> Result<()> {
        // Calculate which conversation history index this corresponds to
        // Each exchange has 2 entries (User: and Claude:)
        let history_index = message_index * 2;
        
        if history_index > self.conversation_history.len() {
            return Err(ClaudeRelayError::Process(
                format!("Invalid undo index: {}", message_index)
            ));
        }
        
        // Save the conversation that will be undone (for restore)
        if history_index < self.conversation_history.len() {
            self.last_undone_history = Some(self.conversation_history.clone());
        }
        
        // Truncate conversation history to this point
        self.conversation_history.truncate(history_index);
        
        // Also truncate conversation states if needed
        self.conversation_states.retain(|state| state.history.len() <= history_index);
        
        Ok(())
    }

    pub fn can_undo(&self) -> bool {
        !self.conversation_states.is_empty()
    }

    pub fn get_last_exchange(&self) -> Result<(String, String)> {
        if self.conversation_history.len() < 2 {
            return Err(ClaudeRelayError::Process("No complete exchange to return".into()));
        }
        
        let user_msg = &self.conversation_history[self.conversation_history.len() - 2];
        let claude_msg = &self.conversation_history[self.conversation_history.len() - 1];
        
        let user_msg = user_msg.strip_prefix("User: ").unwrap_or(user_msg);
        let claude_msg = claude_msg.strip_prefix("Claude: ").unwrap_or(claude_msg);
        
        Ok((user_msg.to_string(), claude_msg.to_string()))
    }

    pub fn can_restore(&self) -> bool {
        self.last_undone_history.as_ref()
            .map(|history| history.len() > self.conversation_history.len())
            .unwrap_or(false)
    }

    pub fn restore_last_undo(&mut self) -> Result<Vec<String>> {
        if !self.can_restore() {
            return Err(ClaudeRelayError::Process("Nothing to restore".into()));
        }
        
        let last_undone = self.last_undone_history.as_ref().unwrap();
        
        // Get the messages that will be restored (for client display)
        let restored_messages = last_undone[self.conversation_history.len()..].to_vec();
        
        // Restore the conversation history
        self.conversation_history = last_undone.clone();
        
        // Clear the undo buffer since we've restored it
        self.last_undone_history = None;
        
        // Rebuild conversation states
        self.save_state();
        
        Ok(restored_messages)
    }

    pub fn get_restored_messages_for_client(&self) -> Vec<(String, String)> {
        if !self.can_restore() {
            return Vec::new();
        }
        
        let last_undone = self.last_undone_history.as_ref().unwrap();
        let restored_part = &last_undone[self.conversation_history.len()..];
        
        let mut messages = Vec::new();
        for chunk in restored_part.chunks(2) {
            if chunk.len() == 2 {
                let user_msg = chunk[0].strip_prefix("User: ").unwrap_or(&chunk[0]);
                let claude_msg = chunk[1].strip_prefix("Claude: ").unwrap_or(&chunk[1]);
                messages.push((user_msg.to_string(), claude_msg.to_string()));
            }
        }
        
        messages
    }
}